src/content/en/updates/2018/08/embind.md

```diff
@@ -0,0 +1,256 @@
+
+This is pretty great as we can catch some errors early instead of dealing with
+the — occasionally quite unwieldy — wasm errors.
+
+### Objects
+
+A pattern that a lot of JavaScript APIs use is options objects. It's a nice
+pattern in JavaScript, but extremely tedious to realize in wasm manually. embind
+can help here, too!
+
+I came up with this _incredibly_ useful C++ function that processes your
+strings, and I urgently want to use it on the web. Here is how I did that:
+
+    #include <emscripten/bind.h>
+    #include <algorithm>
+
+    using namespace emscripten;
+
+    struct ProcessMessageOpts {
+      bool reverse;
+      bool exclaim;
+      int repeat;
+    };
+
+    std::string processMessage(std::string message, ProcessMessageOpts opts) {
+      std::string copy = std::string(message);
+      if(opts.reverse) {
+        std::reverse(copy.begin(), copy.end());
+      }
+      if(opts.exclaim) {
+        copy += "!";
+      }
+      std::string acc = std::string("");
+      for(int i = 0; i < opts.repeat; i++) {
+        acc += copy;
+      }
+      return acc;
+    }
+
+    EMSCRIPTEN_BINDINGS(my_module) {
+      value_object<ProcessMessageOpts>("ProcessMessageOpts")
+        .field("reverse", &ProcessMessageOpts::reverse)
+        .field("exclaim", &ProcessMessageOpts::exclaim)
+        .field("repeat", &ProcessMessageOpts::repeat);
+
+      function("processMessage", &processMessage);
+    }
+
+I am defining a struct for the options of my `processMessage` function. In the
+`EMSCRIPTEN_BINDINGS` block, I can use `value_object` to make JavaScript see
+this C++ value as an object. I could also use `value_array` if I preferred to
+use this C++ value as an array. I also bind the `processMessage` function, and
+the rest is embind _magic_. I can now call the `processMessage` function from
+JavaScript without any preluding code:
+
+    console.log(Module.processMessage(
+      "hello world",
+      {
+        reverse: false,
+        exclaim: true,
+        repeat: 3
+      }
+    )); // Prints "hello world!hello world!hello world!"
+
+### Classes
+
+For completeness sake, I should also show you how embind allows you to expose
+entire classes, which brings a lot of synergy with ES6 classes. You can probably
+start to see a pattern by now:
+
+    #include <emscripten/bind.h>
+    #include <algorithm>
+
+    using namespace emscripten;
+
+    class Counter {
+    public:
+      int counter;
+
+      Counter(int init) :
+        counter(init) {
+      }
+
+      void increase() {
+        counter++;
+      }
+
+      int squareCounter() {
+        return counter * counter;
+      }
+    };
+
+    EMSCRIPTEN_BINDINGS(my_module) {
+      class_<Counter>("Counter")
+        .constructor<int>()
+        .function("increase", &Counter::increase)
+        .function("squareCounter", &Counter::squareCounter)
+        .property("counter", &Counter::counter);
+    }
+
+On the JavaScript side, this almost feels like a native class:
```

Comment:
Sidebar: Should there be a complete reference somewhere how to implement all native idioms so that WASM can someday be used for polyfills?

Reply:
I can bring this up with the wasm tooling team :) We are kinda working towards this anyways, but it’s never been made explicit with a list.

Reply:
Not just polyfills. I would think anyone implementing any kind of API would want to use familiar idioms. I would.